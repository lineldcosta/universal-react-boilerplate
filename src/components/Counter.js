/* @flow */
import React from 'react'
import PropTypes from 'prop-types'

export type Props = {
  count: number,
  onIncrement?: () => void,
  onDecrement?: () => void,
  onReset?: () => void
}

const Counter = ({ count, onIncrement, onDecrement, onReset }: Props) => (
  <div>
    <p className='count'>Counter: {count}</p>
    {onIncrement && <button onClick={onIncrement.bind(this)}>Increment</button>}
    {onDecrement && <button onClick={onDecrement.bind(this)}>Decrement</button>}
    {onReset && <button onClick={onReset.bind(this)}>Reset</button>}
  </div>
)

Counter.propTypes = {
  count: PropTypes.number.isRequired,
  onIncrement: PropTypes.func,
  onDecrement: PropTypes.func,
  onReset: PropTypes.func
}

Counter.defaultProps = {
  count: 0
}

export default Counter
