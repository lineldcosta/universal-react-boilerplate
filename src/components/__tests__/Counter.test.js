/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import Counter from '../Counter'

const setup = propOverrides => {
  const props = Object.assign({
    count: 0,
    onIncrement: jest.fn(),
    onDecrement: jest.fn(),
    onReset: jest.fn()
  }, propOverrides)

  const wrapper = shallow(<Counter { ...props } />)

  return {
    props,
    wrapper,
    count: wrapper.find('.count')
  }
}

it('renders correctly', () => {
  const { wrapper } = setup()
  expect(wrapper).toMatchSnapshot()
})

it('displays the correct count', () => {
  const { count } = setup({ count: 1 })
  expect(count.text()).toEqual('Counter: 1')
})

const testClickButton = (button, property) => {
  it(`calls the '${property}' property when '${button}' is clicked`, () => {
    const { props, wrapper } = setup()
    wrapper.find(`button[children="${button}"]`).simulate('click')
    expect(props[property].mock.calls.length).toEqual(1)
  })
}

testClickButton('Increment', 'onIncrement')
testClickButton('Decrement', 'onDecrement')
testClickButton('Reset', 'onReset')

const testRenderButton = (button, property) => {
  it(`doesn't render '${button}' if '${property}' is null`, () => {
    const { wrapper } = setup({ [property]: null })
    const button = wrapper.find(`button[children="${button}"]`)
    expect(button.length).toEqual(0)
  })
}

testRenderButton('Increment', 'onIncrement')
testRenderButton('Decrement', 'onDecrement')
testRenderButton('Reset', 'onReset')
