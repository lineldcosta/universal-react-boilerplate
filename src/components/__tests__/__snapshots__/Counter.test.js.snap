// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <p
        className="count"
    >
        Counter: 
        0
    </p>
    <button
        onClick={[Function]}
    >
        Increment
    </button>
    <button
        onClick={[Function]}
    >
        Decrement
    </button>
    <button
        onClick={[Function]}
    >
        Reset
    </button>
</div>,
  "nodes": Array [
    <div>
      <p
            className="count"
      >
            Counter: 
            0
      </p>
      <button
            onClick={[Function]}
      >
            Increment
      </button>
      <button
            onClick={[Function]}
      >
            Decrement
      </button>
      <button
            onClick={[Function]}
      >
            Reset
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Counter
        count={0}
        onDecrement={[Function]}
        onIncrement={[Function]}
        onReset={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "count": 0,
          "onDecrement": [Function],
          "onIncrement": [Function],
          "onReset": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <p
                    className="count"
          >
                    Counter: 
                    0
          </p>
          <button
                    onClick={[Function]}
          >
                    Increment
          </button>
          <button
                    onClick={[Function]}
          >
                    Decrement
          </button>
          <button
                    onClick={[Function]}
          >
                    Reset
          </button>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <p
                    className="count"
          >
                    Counter: 
                    0
          </p>
          <button
                    onClick={[Function]}
          >
                    Increment
          </button>
          <button
                    onClick={[Function]}
          >
                    Decrement
          </button>
          <button
                    onClick={[Function]}
          >
                    Reset
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Counter
    count={0}
    onDecrement={[Function]}
    onIncrement={[Function]}
    onReset={[Function]}
/>,
}
`;
